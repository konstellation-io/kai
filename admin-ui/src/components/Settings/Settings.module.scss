@import '../../styles/colors';
@import '../../styles/mixins';

.container {
  @include font-button;
  position: relative;
  line-height: 40px;
  height: 40px;
  text-align: center;
  padding: 0 12px;
  cursor: pointer;
  transition: color ease 0.4s;
  color: $font-color-light;
  background-color: $bg-color-dark;
  z-index: 1;

  &:hover {
    color: $font-color-white;

    .arrow {
      border-color: white;
    }
  }

  &:active {
    box-shadow: none;
  }

  &:after {
    position: absolute;
    width: 7px;
    height: 7px;
    transform: rotate(45deg);
    right: 14px;
    top: 14px;
    content: '';
    border: solid 1px transparent;
    border-bottom-color: white;
    border-right-color: white;
    transition: transform ease 0.3s, top ease 0.3s;
  }

  &.is-open:after {
    transform: rotate(45deg) scale(-1);
    top: 17px;
  }
}

.label {
  display: flex;
  flex-direction: row;
  padding: 0 20px;
  user-select: none;
}

.options {
  transition: max-height ease 0.3s;
  overflow: hidden;
  box-shadow: black 0px 4px 7px;
}
