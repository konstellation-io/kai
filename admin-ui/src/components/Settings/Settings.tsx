import { get } from 'lodash';
import React, { useState, useEffect, useRef } from 'react';
import { useHistory } from 'react-router';
import useEndpoint from '../../hooks/useEndpoint';
import { useApolloClient } from '@apollo/react-hooks';
import { ENDPOINT } from '../../constants/application';
import Button, { BUTTON_ALIGN } from '../Button/Button';
import SettingsIcon from '@material-ui/icons/VerifiedUser';
import AuditIcon from '@material-ui/icons/SupervisorAccount';
import LogoutIcon from '@material-ui/icons/ExitToApp';
import cx from 'classnames';
import styles from './Settings.module.scss';
import ROUTE from '../../constants/routes';
import useClickOutsideListener from '../../hooks/useClickOutsideListener';
import useUserAccess from '../../hooks/useUserAccess';
import { AccessLevel } from '../../graphql/client/typeDefs';

const BUTTON_HEIGHT = 40;
const buttonStyle = {
  paddingLeft: '20%'
};

type Props = {
  label: string;
};

function Settings({ label }: Props) {
  const { accessLevel } = useUserAccess();
  const client = useApolloClient();
  const history = useHistory();
  const [opened, setOpened] = useState(false);
  const [logoutResponse, doLogout] = useEndpoint({
    endpoint: ENDPOINT.LOGOUT,
    method: 'POST'
  });
  const ref = useRef(null);
  useClickOutsideListener({ ref, onClickOutside: () => setOpened(false) });

  useEffect(() => {
    if (logoutResponse.complete) {
      if (get(logoutResponse, 'status') === 200) {
        client.writeData({ data: { loggedIn: false } });
        history.push(ROUTE.LOGIN);
      } else {
        console.error(`Error sending logout request.`);
      }
    }
  }, [logoutResponse, history, client]);

  function getBaseProps(label: string) {
    return {
      label: label.toUpperCase(),
      key: `button${label}`,
      align: BUTTON_ALIGN.LEFT,
      style: buttonStyle
    };
  }

  const settingsButton = (
    <Button
      {...getBaseProps('Settings')}
      to={ROUTE.SETTINGS}
      Icon={SettingsIcon}
    />
  );
  const auditButton = (
    <Button {...getBaseProps('Audit')} to={ROUTE.AUDIT} Icon={AuditIcon} />
  );
  const logoutButton = (
    <Button
      {...getBaseProps('Logout')}
      onClick={() => doLogout()}
      Icon={LogoutIcon}
    />
  );

  const buttons: JSX.Element[] = [];
  switch (accessLevel) {
    case AccessLevel.ADMINISTRATOR:
    case AccessLevel.MANAGER:
      buttons.push(settingsButton, auditButton);
    /* falls through */
    case AccessLevel.VIEWER:
    default:
      buttons.push(logoutButton);
  }

  const nButtons = buttons.length;
  const optionsHeight = nButtons * BUTTON_HEIGHT;

  return (
    <div
      className={cx(styles.container, { [styles['is-open']]: opened })}
      onClick={() => setOpened(!opened)}
      data-testid="settingsContainer"
      ref={ref}
    >
      <div className={styles.label} data-testid="settings-label">
        {label}
      </div>
      <div
        className={styles.options}
        style={{ maxHeight: opened ? optionsHeight : 0 }}
        data-testid="settingsContent"
      >
        {buttons}
      </div>
    </div>
  );
}

export default Settings;
