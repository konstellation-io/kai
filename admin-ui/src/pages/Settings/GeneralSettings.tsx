import React, { useEffect } from 'react';
import useInput from '../../hooks/useInput';

import SettingsHeader from './components/SettingsHeader';
import TextInput from '../../components/Form/TextInput/TextInput';
import Button from '../../components/Button/Button';
import HorizontalBar from '../../components/Layout/HorizontalBar/HorizontalBar';
import ErrorMessage from '../../components/ErrorMessage/ErrorMessage';
import Spinner from '../../components/Spinner/Spinner';
import * as CHECK from '../../components/Form/check';

import cx from 'classnames';
import styles from './Settings.module.scss';

import { useMutation, useQuery } from '@apollo/react-hooks';
import {
  GET_EXPIRATION_TIME,
  UPDATE_EXPIRATION_TIME
} from './Settings.graphql';

const MIN_EXPIRATION_DAYS = 1;
const MAX_EXPIRATION_DAYS = 60;

type FormFieldProps = {
  error: string;
  onChange: Function;
  onSubmit: Function;
  defaultValue: any;
};
function FormField({
  error,
  onChange,
  onSubmit,
  defaultValue
}: FormFieldProps) {
  return (
    <div className={styles.formField}>
      <p className={styles.label}>Session Cookie expiration time</p>
      <div className={styles.input}>
        <TextInput
          whiteColor
          onlyNumbers
          label="nÂº days"
          error={error}
          onChange={onChange}
          onSubmit={onSubmit}
          defaultValue={defaultValue}
        />
      </div>
    </div>
  );
}

/**
 * Expiration time must be a number (days) between 1 and 60
 */
function isExpirationInvalid(value: string) {
  return CHECK.getValidationError([
    CHECK.isFieldNotEmpty(value),
    CHECK.isFieldAnInteger(value, true),
    CHECK.isIntegerWithinRange(value, [
      MIN_EXPIRATION_DAYS,
      MAX_EXPIRATION_DAYS
    ])
  ]);
}

function GeneralSettings() {
  const { data, loading, error: queryError } = useQuery(GET_EXPIRATION_TIME);
  const [updateExpirationTime] = useMutation(UPDATE_EXPIRATION_TIME);
  const { value, isValid, onChange, setValue, error: inputError } = useInput(
    '',
    isExpirationInvalid
  );

  // Sets domains data after receiving API response
  useEffect(() => {
    if (data) {
      setValue(data.settings.cookieExpirationTime);
    }
  }, [data, setValue]);

  if (loading) return <Spinner />;
  if (queryError) return <ErrorMessage />;

  function onSubmit() {
    if (isValid()) {
      const input = {
        cookieExpirationTime: parseInt(value)
      };
      updateExpirationTime({ variables: { input } });
    }
  }

  return (
    <>
      <div className={cx(styles.form, styles.generalSettings)}>
        <SettingsHeader
          title="General settings"
          subtitle="Fusce vehicula dolor arcu, sit amet blandit dolor mollis nec. Donec viverra eleifend
            lacus, vitae ullamcorper metus. Sed sollicitudin ipsum quis nunc sollicitudin ultrices.
            Donec euismod scelerisque ligula. Maecenas eu varius risus, eu aliquet arcu. Curabitur
            fermentum suscipit est, tincidunt."
        />
        <FormField
          error={inputError}
          onChange={onChange}
          onSubmit={onSubmit}
          defaultValue={data.settings.cookieExpirationTime}
        />
      </div>
      <HorizontalBar>
        <Button label={'SAVE CHANGES'} primary onClick={onSubmit} />
      </HorizontalBar>
    </>
  );
}

export default GeneralSettings;
