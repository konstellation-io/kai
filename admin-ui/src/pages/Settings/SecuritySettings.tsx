import { get } from 'lodash';

import React, { useState, useEffect } from 'react';
import useInput from '../../hooks/useInput';

import DomainIcon from '@material-ui/icons/Language';

import SettingsHeader from './components/SettingsHeader';
import TextInput from '../../components/Form/TextInput/TextInput';
import Button from '../../components/Button/Button';
import DomainList from '../../components/DomainList/DomainList';
import * as CHECK from '../../components/Form/check';

import cx from 'classnames';
import styles from './Settings.module.scss';

import { useMutation, useQuery } from '@apollo/react-hooks';
import {
  GET_DOMAINS,
  UPDATE_DOMAINS,
  SettingsResponse,
  SettingsVars
} from './Settings.graphql';

type FormFieldProps = {
  value?: string;
  error: string;
  onChange: Function;
  onSubmit: Function;
};

function FormField({ value, error, onChange, onSubmit }: FormFieldProps) {
  return (
    <div className={styles.formField}>
      <DomainIcon className="icon-regular" />
      <p className={styles.label}>Domain white list</p>
      <div className={styles.input}>
        <TextInput
          whiteColor
          label="domain name"
          error={error}
          onChange={onChange}
          onSubmit={onSubmit}
          formValue={value}
        />
      </div>
      <div className={styles.button}>
        <Button label={'ADD DOMAIN'} onClick={onSubmit} border />
      </div>
    </div>
  );
}

function validateForm(value: string, domains: string[]) {
  return CHECK.getValidationError([
    CHECK.isFieldNotEmpty(value),
    CHECK.isFieldAnString(value),
    CHECK.isDomainValid(value),
    domainDuplicated(value, domains)
  ]);
}

function domainDuplicated(newDomain: string, domains: string[]) {
  const valid = !domains.includes(newDomain);
  const msg = valid ? '' : 'Duplicated domain';
  return { valid, message: msg };
}

function SecuritySettings() {
  const [allowedDomains, setAllowedDomains] = useState([]);

  const { data: queryData, loading, error: queryError } = useQuery<
    SettingsResponse
  >(GET_DOMAINS, { fetchPolicy: 'no-cache' });
  const [updateAllowedDomain] = useMutation<SettingsResponse, SettingsVars>(
    UPDATE_DOMAINS,
    { onCompleted: onCompleteUpdateDomain }
  );
  const {
    value,
    isValid,
    onChange,
    error: inputError,
    clear,
    clearError
  } = useInput('', (val: string) => validateForm(val, allowedDomains));

  // Set domains data after retrieving it from API
  useEffect(() => {
    if (get(queryData, 'settings')) {
      setAllowedDomains(get(queryData, 'settings.authAllowedDomains'));
    }
  }, [queryData]);
  // Set domains data after making a mutation
  function onCompleteUpdateDomain(updatedData: any) {
    setAllowedDomains(updatedData.updateSettings.settings.authAllowedDomains);
  }

  function updateDomains(newDomains: any) {
    const input = { authAllowedDomains: newDomains };
    updateAllowedDomain({ variables: { input } });
    setAllowedDomains(newDomains);

    clear();
  }

  function onSubmit() {
    if (isValid()) {
      const domain = value.toLowerCase();
      console.log(`Domain ${domain} added`);

      const newDomains = allowedDomains.concat(domain);
      updateDomains(newDomains);
      clear();
      clearError();
    }
  }

  function onRemoveDomain(domain: string) {
    console.log(`Domain ${domain} removed`, allowedDomains, domain);
    const newDomains = allowedDomains.filter(d => d !== domain);
    updateDomains(newDomains);
    clearError();
  }

  function getContent() {
    return (
      <DomainList
        onRemoveDomain={onRemoveDomain}
        error={queryError}
        loading={loading}
        data={allowedDomains}
      />
    );
  }

  return (
    <>
      <div className={cx(styles.form, styles.securitySettings)}>
        <SettingsHeader
          title="Security settings"
          subtitle="Fusce vehicula dolor arcu, sit amet blandit dolor mollis nec. Donec viverra eleifend
            lacus, vitae ullamcorper metus. Sed sollicitudin ipsum quis nunc sollicitudin ultrices.
            Donec euismod scelerisque ligula. Maecenas eu varius risus, eu aliquet arcu. Curabitur
            fermentum suscipit est, tincidunt."
        />
        <FormField
          error={inputError}
          onChange={onChange}
          onSubmit={onSubmit}
          value={value}
        />
        <div className={styles.domains}>{getContent()}</div>
      </div>
    </>
  );
}

export default SecuritySettings;
