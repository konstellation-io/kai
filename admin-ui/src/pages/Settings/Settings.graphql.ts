import gql from 'graphql-tag';
import { Settings } from '../../graphql/models';
export interface SettingsResponse {
  settings: Settings;
}

export interface SettingsVars {
  input: Settings;
}

export const GET_DOMAINS = gql`
  query GetDomains {
    settings {
      authAllowedDomains
    }
  }
`;

export const GET_EXPIRATION_TIME = gql`
  query GetSettings {
    settings {
      sessionLifetimeInDays
    }
  }
`;

export const UPDATE_SESSION_LIFETIME = gql`
  mutation UpdateSettings($input: SettingsInput!) {
    updateSettings(input: $input) {
      errors {
        code
        message
      }
      settings {
        sessionLifetimeInDays
      }
    }
  }
`;

export const UPDATE_DOMAINS = gql`
  mutation UpdateDomains($input: SettingsInput!) {
    updateSettings(input: $input) {
      errors {
        code
        message
      }
      settings {
        authAllowedDomains
      }
    }
  }
`;
