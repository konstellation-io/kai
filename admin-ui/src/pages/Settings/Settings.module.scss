@import '../../styles/colors';
@import '../../styles/mixins';

.container {
  height: 100%;
  display: flex;
  flex-direction: row;
  width: 100%;
}

.content {
  height: 100%;
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  overflow: auto;
}

.form {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  margin: 40px 40px 0 85px;
  min-width: 540px;
  overflow: hidden;
}
.generalSettings {
  .formField {
    height: 140px;
    display: flex;
    flex-direction: row;
    align-items: center;
    margin-top: 30px;
    border: solid $bg-color-lighter;
    border-width: 2px 0;

    .label {
      @include font-body;
      color: font-color(light);

      margin: 0 40px 0 30px;
    }

    svg {
      margin-left: 20px;
    }

    .input {
      width: 140px;
    }
  }
}

.securitySettings {
  margin: 40px 40px 75px 85px;
}

.formTitle {
  @include font-subheader;
  color: font-color(light);
}

.formDescription {
  @include font-body;

  margin-top: 10px;
}

.listsContainer {
  display: flex;
  overflow: hidden;
  height: 100vh;
  & > div:first-child {
    margin-right: 48px;
  }
}

@media (max-width: 1400px) {
  .listsContainer {
    flex-direction: column;
    & div:first-child {
      margin-right: 0;
    }
  }
}
