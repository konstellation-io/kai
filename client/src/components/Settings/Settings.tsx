import React, {useState} from 'react';

import Button, {BUTTON_TYPES, BUTTON_ALIGN} from '../Button/Button';
import LogoutIcon from '@material-ui/icons/ExitToApp';

import styles from './Settings.module.scss';


const BUTTON_HEIGHT = 40;
const buttonStyle = {
  paddingLeft: '20%'
};

function doLogout() {
  console.log('LOGGED OUT');
}

type Props = {
  label: string;
};

function Settings({ label }: Props) {
  const [opened, setOpened] = useState(false);

  const buttons = [
    <Button
      label={'LOGOUT'}
      type={BUTTON_TYPES.GREY}
      onClick={doLogout}
      Icon={LogoutIcon}
      align={BUTTON_ALIGN.LEFT}
      style={buttonStyle}
      key={'buttonLogout'}
    />
  ];
  const nButtons = buttons.length;
  const optionsHeight = nButtons * BUTTON_HEIGHT;

  return (
    <div
      className={styles.container}
      onMouseEnter={ () => setOpened(true) }
      onMouseLeave={ () => setOpened(false) }
      data-testid='settingsContainer'
    >
      <div className={styles.label}>
        { label }
        <div className={styles.arrow} />
      </div>
      <div
        className={styles.options}
        style={{ maxHeight: opened ? optionsHeight : 0 }}
        data-testid='settingsContent'
      >
        { buttons }
      </div>
    </div>
  );
}

export default Settings;
