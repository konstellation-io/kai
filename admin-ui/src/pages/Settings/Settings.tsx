import React from 'react';
import { Route } from 'react-router';
import ROUTE from '../../constants/routes';

import GeneralIcon from '@material-ui/icons/DeviceHub';
import SecurityIcon from '@material-ui/icons/Security';

import GeneralSettings from './GeneralSettings';
import SecuritySettings from './SecuritySettings';
import Sidebar from '../../components/Sidebar/Sidebar';

import styles from './Settings.module.scss';
import PageBase from '../../components/Layout/PageBase/PageBase';

const tabs = [
  {
    label: 'GENERAL',
    route: ROUTE.SETTINGS_GENERAL,
    Icon: GeneralIcon
  },
  {
    label: 'SECURITY',
    route: ROUTE.SETTINGS_SECURITY,
    Icon: SecurityIcon
  }
];

function Settings() {
  return (
    <PageBase>
      <div className={styles.container} data-testid="settingsContainer">
        <Sidebar title="Settings" tabs={tabs} />
        <div className={styles.content}>
          <Route
            exact
            path={ROUTE.SETTINGS_GENERAL}
            component={GeneralSettings}
          />
          <Route
            exact
            path={ROUTE.SETTINGS_SECURITY}
            component={SecuritySettings}
          />
        </div>
      </div>
    </PageBase>
  );
}

export default Settings;
