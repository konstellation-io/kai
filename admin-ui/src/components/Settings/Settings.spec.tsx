import React from 'react';
import { MemoryRouter } from 'react-router';
import { render, fireEvent, cleanup } from '@testing-library/react';
import Settings from './Settings';
import { getByTestId } from '@testing-library/dom';
import '@testing-library/jest-dom/extend-expect';

afterEach(cleanup);

function renderComponent() {
  return render(
    <MemoryRouter>
      <Settings />
    </MemoryRouter>
  );
}

it('Renders Settings without crashing', () => {
  const { container } = renderComponent();
  expect(container).toMatchSnapshot();
});

it('Shows logout option', () => {
  const { getByText } = renderComponent();

  expect(getByText('LOGOUT')).toBeInTheDocument();
});

it('Shows options on mouse enter', () => {
  const { container } = renderComponent();

  const settingsContent = getByTestId(container, 'settingsContent');
  // @ts-ignore
  expect(settingsContent.style['max-height']).toBe('0');

  fireEvent.mouseEnter(getByTestId(container, 'settingsContainer'));
  // @ts-ignore
  expect(settingsContent.style['max-height']).not.toBe('0');

  fireEvent.mouseLeave(getByTestId(container, 'settingsContainer'));
  // @ts-ignore
  expect(settingsContent.style['max-height']).toBe('0');
});

//TODO: make logout test
