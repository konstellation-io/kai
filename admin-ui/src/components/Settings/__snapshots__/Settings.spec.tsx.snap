// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Settings without crashing 1`] = `
<div>
  <div
    class="container"
    data-testid="settingsContainer"
  >
    <div
      class="label"
    />
    <div
      class="options"
      data-testid="settingsContent"
      style="max-height: 0;"
    >
      <a
        data-testid="buttonLink"
        href="/settings"
      >
        <div
          class="btn grey left label"
          style="padding-left: 20%; height: 40px; line-height: 40px;"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root icon-regular"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 1L3 5v6c0 5.55 3.84 10.74 9 12 5.16-1.26 9-6.45 9-12V5l-9-4zm-2 16l-4-4 1.41-1.41L10 14.17l6.59-6.59L18 9l-8 8z"
            />
          </svg>
          <span>
            SETTINGS
          </span>
        </div>
      </a>
      <a
        data-testid="buttonLink"
        href="/audit"
      >
        <div
          class="btn grey left label"
          style="padding-left: 20%; height: 40px; line-height: 40px;"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root icon-regular"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.5 12c1.38 0 2.49-1.12 2.49-2.5S17.88 7 16.5 7C15.12 7 14 8.12 14 9.5s1.12 2.5 2.5 2.5zM9 11c1.66 0 2.99-1.34 2.99-3S10.66 5 9 5C7.34 5 6 6.34 6 8s1.34 3 3 3zm7.5 3c-1.83 0-5.5.92-5.5 2.75V19h11v-2.25c0-1.83-3.67-2.75-5.5-2.75zM9 13c-2.33 0-7 1.17-7 3.5V19h7v-2.25c0-.85.33-2.34 2.37-3.47C10.5 13.1 9.66 13 9 13z"
            />
          </svg>
          <span>
            AUDIT
          </span>
        </div>
      </a>
      <div
        class="btn grey left label"
        style="padding-left: 20%; height: 40px; line-height: 40px;"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon-regular"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span>
          LOGOUT
        </span>
      </div>
    </div>
  </div>
</div>
`;
